sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala

```diff
@@ -260,6 +260,20 @@ private[engine] final class Preprocessor(
     ) {
       commandPreprocessor.unsafePreprocessInterfaceView(templateId, argument, interfaceId)
     }
+
+  def prefetchKeys(
+      commands: ImmArray[speedy.Command]
+  ): Result[Unit] = {
+    val keys = commands.iterator.collect {
+      case speedy.Command.ExerciseByKey(templateId, contractKey, _, _) =>
+        val packageTxVersion = speedy.Speedy.Machine.tmplId2TxVersion(compiledPackages, templateId)
+        val pkgName =
+          speedy.Speedy.Machine.tmplId2PackageName(compiledPackages, templateId, packageTxVersion)
+        speedy.Speedy.Machine.toGlobalKey(packageTxVersion, pkgName, templateId, contractKey)
```

Comment:
I would push the computation of packageTxVersion and pkgName inside toGlobalKey.
This probably needs a bit of refactoring.
Feel free to ignore if you prefer not refactoring.